nifi-nar-bundles/nifi-metrics-reporting-bundle/nifi-metrics-reporting-task/src/main/java/org/apache/nifi/metrics/reporting/reporter/service/GraphiteMetricReporterService.java

```diff
@@ -0,0 +1,147 @@
+            .build();
+
+    /** Points to the port on which the graphite server listens. */
+    public static final PropertyDescriptor PORT = new PropertyDescriptor.Builder()
+            .name("port")
+            .displayName("port")
+            .description("The port on which carbon listens")
+            .required(true)
+            .addValidator(StandardValidators.PORT_VALIDATOR)
+            .build();
+
+    /** Points to the charset name that the graphite server expects. */
+    public static final PropertyDescriptor CHARSET = new PropertyDescriptor.Builder()
+            .name("charset")
+            .displayName("charset")
+            .description("The charset used by the graphite server")
+            .required(true)
+            .defaultValue("UTF-8")
+            .addValidator(StandardValidators.CHARACTER_SET_VALIDATOR)
+            .build();
+
+    /** List of property descriptors used by the service. */
+    private static final List<PropertyDescriptor> properties;
+
+    static {
+        final List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(HOST);
+        props.add(PORT);
+        properties = Collections.unmodifiableList(props);
+    }
+
+    /** Graphite sender, a connection to the server. */
+    private GraphiteSender graphiteSender;
+
+    /**
+     * Create the {@link #graphiteSender} according to configuration.
+     *
+     * @param context used to access properties.
+     */
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        String host = context.getProperty(HOST).getValue();
+        int port = context.getProperty(PORT).asInteger();
+        Charset charset = Charset.forName(context.getProperty(CHARSET).getValue());
+        graphiteSender = createSender(host, port, charset);
+    }
+
+    /**
+     * Close the graphite sender.
+     *
+     * @throws IOException if failed to close the connection.
+     */
+    @OnDisabled
+    public void shutdown() throws IOException {
+        try {
+            graphiteSender.close();
+        } finally {
+            graphiteSender = null;
+        }
+    }
+
+    /**
+     * Use the {@link #graphiteSender} in order to create a reporter.
+     *
+     * @param metricRegistry registry with the metrics to report.
+     * @return a reporter instance.
+     */
+    @Override
+    public ScheduledReporter createReporter(MetricRegistry metricRegistry) {
+        return GraphiteReporter.forRegistry(metricRegistry).build(graphiteSender);
+
+    }
+
+    /**
+     * Create a sender.
+     *
+     * @param host the hostname of the server to connect to.
+     * @param port the port on which the server listens.
+     * @param charset the charset in which the server expects logs.
+     * @return The created sender.
+     */
+    protected GraphiteSender createSender(String host, int port, Charset charset) {
+        return new Graphite(host, port, SocketFactory.getDefault(), charset);
```

Comment:
Is there ever a case where the metrics would be sent over an SSL/TLS connection?

Reply:
I am not a graphite expert, but as far as I know there is no native support for that - as graphite is meant to be used in a trusted (or secure) network. It might be possible to communicate with carbon using rabbitmq securely in some way, since it supports amqp and there is a rabbit mq client - but I feel that it is out of scope for this PR (correct me if I am wrong and I'll be happy to look into it).

Also if someone can correct me on this and knows of a proper way to feed carbon with data in a secure fashion I'd be more than happy to stand corrected.